database/src/ops/blockchain.rs

```diff
@@ -1,9 +1,124 @@
 //! Blockchain.
 
 //---------------------------------------------------------------------------------------------------- Import
+use monero_serai::transaction::Timelock;
+
+use cuprate_types::VerifiedBlockInformation;
+
+use crate::{
+    database::{DatabaseRo, DatabaseRw},
+    env::EnvInner,
+    error::RuntimeError,
+    ops::macros::doc_error,
+    tables::{
+        BlockBlobs, BlockHeights, BlockInfos, KeyImages, NumOutputs, Outputs, PrunableHashes,
+        PrunableTxBlobs, PrunedTxBlobs, RctOutputs, Tables, TablesMut, TxHeights, TxIds,
+        TxUnlockTime,
+    },
+    transaction::{TxRo, TxRw},
+    types::{BlockHash, BlockHeight, BlockInfo, KeyImage, Output, PreRctOutputId, RctOutput},
+};
 
 //---------------------------------------------------------------------------------------------------- Free Functions
-/// TODO
-pub fn height() {
-    todo!()
+/// Retrieve the height of the chain.
+///
+/// This returns the chain-tip, not the [`top_block_height`].
+///
+/// For example:
+/// - The blockchain has 0 blocks => this returns `0`
+/// - The blockchain has 1 block (height 0) => this returns `1`
+/// - The blockchain has 2 blocks (height 1) => this returns `2`
+///
+/// So the height of a new block would be `chain_height()`.
+#[doc = doc_error!()]
+#[inline]
+pub fn chain_height(
+    table_block_heights: &impl DatabaseRo<BlockHeights>,
+) -> Result<BlockHeight, RuntimeError> {
+    table_block_heights.len()
+}
+
+/// Retrieve the height of the top block.
+///
+/// This returns the height of the top block, not the [`chain_height`].
+///
+/// For example:
+/// - The blockchain has 0 blocks => this returns `Err(RuntimeError::KeyNotFound)`
+/// - The blockchain has 1 block (height 0) => this returns `Ok(0)`
+/// - The blockchain has 2 blocks (height 1) => this returns `Ok(1)`
+#[doc = doc_error!()]
+#[inline]
+pub fn top_block_height(
+    table_block_heights: &impl DatabaseRo<BlockHeights>,
+) -> Result<BlockHeight, RuntimeError> {
+    match table_block_heights.len()? {
+        0 => Err(RuntimeError::KeyNotFound),
```

Comment:
At which point could the blockchain has 0 blocks ?

Reply:
Before initial sync? Unless we are supposed to special case the first block manually?

Reply:
> Unless we are supposed to special case the first block manually?

That's what I'm wondering. Genesis block is a constant. We're still supposed to receive it from network iirc, it should be accepted without questioning, so impossible we hit 0. Maybe I'm wrong on my thinking. @Boog900 ?

Reply:
Another part of Cuprate will handle the genesis block, it is hard coded but different for each network

Reply:
Alright so impossible to get 0. You can make it `unreachable!()`

Reply:
Although, is there a reason why we should panic? Returning `Result` where possible over panic is preferred IMO - the function already returns `Result`.

I don't think it'll do anything for optimizations since `0` is always branched for:
```rust
match table_block_heights.len()? {
    0 => panic!(), // branch
    height => Ok(height - 1),
}

// or...
//          this operator branches and panics too
//                         v
table_block_heights.len()? - 1
```
Thinking about non-`cuprated` use-cases too.

Reply:
> Although, is there a reason why we should panic?

Just to avoid the overhead of handling that case higher in block downloader

>  Returning Result where possible over panic is preferred IMO. Thinking about non-cuprated use-cases too.

Fair enough. The error is a little unclear so maybe add a mention in the doc "If no blocks registered yet, then this function return KeyNotFound"